File: numerics/visualization/plotStats.r
#!/usr/bin/Rscript --vanilla

# plot means and standard deviations
# indices is a vector of species indices to plot

# Example to run the function from $STOCHKIT_HOME after running dimer decay example
# plotStats.r models/examples/dimer_decay_output/stats 1 2 3
# where "1 2 3" specifies printing species indexes 1,2 and 3

# Process command line arguments
args.vec <- commandArgs(TRUE)
if(length(args.vec) < 2) { stop("Usage: plotStats.r inputFileDirectory(CHARACTER) outputFile(CHARACTER) [indices](NUMERIC)") }
inputFileDirectory <- args.vec[1]
outputFile <- args.vec[2]
if(length(args.vec) >= 3) {
  indices <- as.numeric(args.vec[3:length(args.vec)])
} else {
  indices <- NULL
}

# Color vector
colors <- c("blue","red","green","magenta","cyan","black","yellow","lightgreen","purple","deeppink")

# Build full inpute file names
fnameMean <- paste(inputFileDirectory,"means.txt",sep="/")
fnameVar <- paste(inputFileDirectory,"variances.txt",sep="/")

# Read in the means.txt data
mheader <- read.table(fnameMean,as.is=TRUE,nrows=1)
if(any(sapply(mheader,is.character))) {
  mheader <- as.character(mheader)
  mdata <- as.matrix(read.table(fnameMean,as.is=TRUE,skip=1))
} else {
  mheader <- ""
  mdata <- as.matrix(read.table(fnameMean,as.is=TRUE))
}

# Read in the variances.txt data
vheader <- read.table(fnameVar,as.is=TRUE,nrows=1)
if(any(sapply(vheader,is.character))) {
  vheader <- as.character(vheader)
  vdata <- as.matrix(read.table(fnameVar,as.is=TRUE,skip=1))
} else {
  vheader <- ""
  vdata <- as.matrix(read.table(fnameVar,as.is=TRUE))
}

# Check and process indices
if(!is.null(indices)) {
  # Check to make sure that the indices are integers
  if(!all.equal(as.integer(indices), indices)) { stop("need integer indices") }

  # Check to make sure that the indices are greater than 0
  if(any(indices <= 0)) { stop("indices need to be positive integers") }

  # Check to eliminate duplicate indices
  if(any(isDuplicated <- duplicated(indices))) {
    cat(paste("removing ",sum(isDuplicated)," duplicate indices...\n",sep=""))
    indices <- unique(indices)
  }

  # Add 1 to account for the time column; sort indices
  indices <- indices + 1
  indices <- sort(indices)

  # Check to make sure indices are within range
  if(indices[length(indices)] > ncol(mdata)) { stop(paste("indices must be less than or equal to ",ncol(mdata)-1,sep="")) }
} else {
  # Populate indices if empty
  indices <- 2:ncol(mdata)
}
len_indices <- length(indices)

# Check equality of mean and variance headers
if(!all(mheader == vheader)) { stop("ERROR: mean and variance should have identical labels") }

# Extract and check equality of mean and variance time points
mtimes <- mdata[,1]
vtimes <- vdata[,1]
if(!all(mtimes == vtimes)) { stop("ERROR: mean and variance should have the same time points") }

# Extract means, variances, and upper/lower confidence intervals
means <- mdata[,indices,drop=FALSE]
vars <- vdata[,indices,drop=FALSE]
mupper <- means + sqrt(vars)
mlower <- means - sqrt(vars)

# Create plot labels
if((length(mheader)==1) && (mheader=="")) {
  labels <- paste("Index",indices-1,sep=" ")
} else {
  labels <- mheader[indices]
}

# Make plots
png(outputFile,width=1200,height=900)
par(xaxs="i",yaxs="i",oma=c(2.5,4,1,3),mar=c(4,7,4,5))
if(length(mtimes) == 1) {
  xlim <- c(mtimes[1]*.9, mtimes[1]*1.1)
} else {
  xlim <- c(mtimes[1], mtimes[length(mtimes)])
}
if(len_indices > length(colors)) { colors <- c(colors,sample(colors(),len_indices-length(colors))) }
plot(NA, NA, xlim=xlim, ylim=range(c(mupper,mlower)), xlab="", ylab="", main="stats plot", cex.main=2.25, axes=FALSE)
axis(1,tck=.01,cex.axis=1.75)
axis(2,tck=.01,cex.axis=1.75,las=2,hadj=.8)
axis(3,labels=FALSE,tck=.01)
axis(4,labels=FALSE,tck=.01)
box()
mtext(side=1,text="time",cex=2.25,line=3)
mtext(side=2,text="population",cex=2.25,line=5)
for(i in 1:len_indices) {
  points(mtimes, means[,i], col=colors[i], type="l", lty=1, lwd=4)
  points(mtimes, mupper[,i], col=colors[i], type="l", lty=2, lwd=2)
  points(mtimes, mlower[,i], col=colors[i], type="l", lty=2, lwd=2)
}
legend("topright",legend=labels,lty=1,col=colors[1:len_indices],lwd=4,cex=2)
graphics.off()

File: google_appengine/helloworld/StochKit2.0.6/tools/MATLAB/plotStats.m
function plotStats(inputFileDirectory,indices)
clc;
%  plot means and standard deviations
% variances fiels are stored
% indices is a vector of species indices to plot
% label is a integer where label=0 means the means and variances files do
% not contain labels. label=1 means the first line of the file contains
% labels for species. label=1 will cause the labels to be displayed in the
% legend of the plot. label=0 will use species indices as legend.

% Example to run the function from $STOCHKIT_HOME/tools/MATLAB after running dimer decay example
% plotStats('../../models/examples/dimer_decay_output/stats',[1,2,3])
% where [1,2,3] specifies printing species indexes 1,2 and 3

% Color vector
colors = [  0 0 1; % blue
    1 0 0;  % red
    0 1 0; % green
    1 0 1; % magenta
    0 1 1;  % cyan
    0 0 0; % black
    1 1 0;  % yellow
    0.6377    0.9577    0.2407; % light green
    0.8754    0.5181    0.9436; % purple
    0.6951    0.0680    0.2548; % dark pink
    ];

% Get time data from the first input file and use it as a reference
fnameMean = fullfile(inputFileDirectory,'means.txt');
fnameVar =  fullfile(inputFileDirectory,'variances.txt');


% Check to make sure that the indices are integers
if (sum((round(indices)==indices)) ~= length(indices))
    error('need integer indices')
end

% Check to make sure that the indices are greater than 0
if (min(indices)<=0)
    error('indices need to be positive integers')
end

% Check to eliminate duplicate indices
[indices2, m, n] =unique(indices);
if (length(indices2)~=length(indices))
    countDuplicate = length(n) - length(m);
    display(['removing ',num2str(countDuplicate),' duplicate indices...'])
    indices=indices2;
end

indices = indices+1; % add 1 to account for the time column
indices = sort(indices);
len_indices =length(indices);


fid1 = fopen(fnameMean);
if fid1==-1
  error('means.txt file not found or permission denied');
end
fid2 = fopen(fnameVar);
if fid2==-1
  error('variances.txt file not found or permission denied');
end


% Read in the means.txt data
no_lines = 0;
max_line = 0;
ncols = 0;
mdata = [];

line = fgetl(fid1);
if ~ischar(line)
    error('ERROR: means.txt file contains no data')
end
[mdata, ncols, errmsg, nxtindex] = sscanf(line, '%f');

while isempty(mdata)||(nxtindex==1)
    no_lines = no_lines+1;
    max_line = max([max_line, length(line)]);
    eval(['line', num2str(no_lines), '=line;']);
    line = fgetl(fid1);
  if ~ischar(line)
      error('ERROR: file contains no data')
  end
  [mdata, ncols, errmsg, nxtindex] = sscanf(line, '%f');
end 

mdata = [mdata; fscanf(fid1, '%f')];
fclose(fid1);


mheader = setstr(' '*ones(no_lines, max_line));
for i = 1:no_lines
    varname = ['line' num2str(i)];
    if eval(['length(' varname ')~=0'])
        eval(['mheader(i, 1:length(' varname ')) = ' varname ';'])
    end
end 

eval('mdata = reshape(mdata, ncols, length(mdata)/ncols)'';', '');



% Read in the variances.txt data
no_lines = 0;
max_line = 0;
ncols = 0;
vdata = [];

line = fgetl(fid2);
if ~ischar(line)
    error('ERROR: means.txt file contains no data')
end
[vdata, ncols, errmsg, nxtindex] = sscanf(line, '%f');

while isempty(vdata)||(nxtindex==1)
    no_lines = no_lines+1;
    max_line = max([max_line, length(line)]);
    eval(['line', num2str(no_lines), '=line;']);
    line = fgetl(fid2);
  if ~ischar(line)
      error('ERROR: file contains no data')
  end
  [vdata, ncols, errmsg, nxtindex] = sscanf(line, '%f');
end 
vdata = [vdata; fscanf(fid2, '%f')];
fclose(fid2);

vheader = setstr(' '*ones(no_lines, max_line));
for i = 1:no_lines
    varname = ['line' num2str(i)];
    if eval(['length(' varname ')~=0'])
        eval(['vheader(i, 1:length(' varname ')) = ' varname ';'])
    end
end 

eval('vdata = reshape(vdata, ncols, length(vdata)/ncols)'';', '');


mtimes = mdata(:,1);
vtimes = vdata(:,1);
if ~isequal(mtimes, vtimes)
    error('ERROR: mean and variance should have the same time points');
end

if ~strcmp(mheader, vheader)
    error('ERROR: mean and variance should have identical labels');
end

means = mdata(:,indices);
vars = vdata(:,indices);
inds= regexpi(mheader,'\s');

if strcmp(mheader,'')
    labels= cell(1,length(indices));
    for i=1:length(indices)
        labels(i)=cellstr(['Index ',num2str(indices(i)-1)]);
    end
else
    labels = cell(1,len_indices);
    for i=1:(len_indices-1)
        labels{i} = mheader(inds(indices(i)-1)+1:inds(indices(i))-1);
    end
    if indices(end)==ncols
        labels{len_indices} = mheader(inds(ncols-1)+1:end);
    else
        labels{len_indices} = mheader(inds(indices(end)-1)+1:inds(indices(end))-1);
    end
end


if length(indices)<length(colors)
    for i=1:length(indices)
        plot(mtimes, means(:,i),'Color',colors(i,:),'LineWidth',2);
        hold on
        plot(mtimes,means(:,i)+sqrt(vars(:,i)),'--','Color',colors(i,:),'HandleVisibility','off')
        plot(mtimes,means(:,i)-sqrt(vars(:,i)),'--','Color',colors(i,:),'HandleVisibility','off')
    end
else
    for i=1:length(colors)
        plot(mtimes,means(:,i),'Color',colors(i,:),'LineWidth',2)
        hold on
        plot(mtimes,means(:,i)+sqrt(vars(:,i)),'--','Color',colors(i,:),'HandleVisibility','off')
        plot(mtimes,means(:,i)-sqrt(vars(:,i)),'--','Color',colors(i,:),'HandleVisibility','off')
    end
    for i=length(colors)+1:length(indices)
        randomColor = [rand rand rand];
        plot(mtimes,means(:,i),'Color',randomColor,'LineWidth',2)
        hold on
        plot(mtimes,means(:,i)+sqrt(vars(:,i)),'--','Color',randomColor,'HandleVisibility','off')
        plot(mtimes,means(:,i)-sqrt(vars(:,i)),'--','Color',randomColor,'HandleVisibility','off')
    end
end
hold off
if (length(mtimes)==1)
    xlim([mtimes(1)*.9 mtimes(1)*1.1])
else
    xlim([mtimes(1),mtimes(end)])  
end
legend(labels)
xlabel('time','FontSize',14,'FontWeight','demi')
ylabel('population','FontSize',14,'FontWeight','demi')
title('stats plot','FontSize',14,'FontWeight','demi')
